cost_estimate.go

```diff
@@ -0,0 +1,129 @@
+
+	// Logs retrieves the logs of a costEstimate.
+	Logs(ctx context.Context, costEstimateID string) (io.Reader, error)
+}
+
+// costEstimates implements CostEstimates.
+type costEstimates struct {
+	client *Client
+}
+
+// CostEstimateStatus represents a costEstimate state.
+type CostEstimateStatus string
+
+//List all available costEstimate statuses.
+const (
+	CostEstimateCanceled CostEstimateStatus = "canceled"
+	CostEstimateErrored  CostEstimateStatus = "errored"
+	CostEstimateFinished CostEstimateStatus = "finished"
+	CostEstimateQueued   CostEstimateStatus = "queued"
+)
+
+// CostEstimate represents a Terraform Enterprise costEstimate.
+type CostEstimate struct {
+	ID           string `jsonapi:"primary,cost-estimations"`
+	ErrorMessage string `jsonapi:"attr,error-message"`
+
+	ResourcesCount          string `jsonapi:"attr,resources-count"`
+	MatchedResourcesCount   string `jsonapi:"attr,matched-resources-count"`
+	UnmatchedResourcesCount string `jsonapi:"attr,unmatched-resources-count"`
+	PriorMonthlyCost        string `jsonapi:"attr,prior-monthly-cost"`
+	ProposedMonthlyCost     string `jsonapi:"attr,proposed-monthly-cost"`
+	DeltaMonthlyCost        string `jsonapi:"attr,delta-monthly-cost"`
+
+	Status           CostEstimateStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *CostEstimateStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// CostEstimateStatusTimestamps holds the timestamps for individual costEstimate statuses.
+type CostEstimateStatusTimestamps struct {
+	CanceledAt time.Time `json:"canceled-at"`
+	ErroredAt  time.Time `json:"errored-at"`
+	FinishedAt time.Time `json:"finished-at"`
+	QueuedAt   time.Time `json:"queued-at"`
+}
+
+// Read a costEstimate by its ID.
+func (s *costEstimates) Read(ctx context.Context, costEstimateID string) (*CostEstimate, error) {
+	if !validStringID(&costEstimateID) {
+		return nil, errors.New("invalid value for cost estimation ID")
+	}
+
+	u := fmt.Sprintf("cost-estimations/%s", url.QueryEscape(costEstimateID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ce := &CostEstimate{}
+	err = s.client.do(ctx, req, ce)
+	if err != nil {
+		return nil, err
+	}
+
+	return ce, nil
+}
+
+// Logs retrieves the logs of a costEstimate.
+func (s *costEstimates) Logs(ctx context.Context, costEstimateID string) (io.Reader, error) {
+	if !validStringID(&costEstimateID) {
+		return nil, errors.New("invalid value for cost estimation ID")
+	}
+
+	// Loop until the context is canceled or the cost estimation is finished
+	// running. The cost estimation logs are not streamed and so only available
+	// once the estimation is finished.
+	for {
+		// Get the costEstimate to make sure it exists.
+		ce, err := s.Read(ctx, costEstimateID)
+		if err != nil {
+			return nil, err
+		}
+
+		switch ce.Status {
+		case CostEstimateQueued:
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			case <-time.After(500 * time.Millisecond):
```

Comment:
I think consistently polling every 500ms may be a bit too aggressive for the API. Maybe we should implement something like this instead: https://github.com/hashicorp/go-tfe/blob/master/logreader.go#L47

Reply:
I agree that 500 feels harsh but this worker is pretty quick and shouldn't take too many iterations to return. 500 matches the policy check interval, at least.